specs/RainbowBridge/README.md

```diff
@@ -0,0 +1,388 @@
+
+For NEAR we've selected **cryptographic proofs of transaction execution results** on the NEAR network.  This is called `NearTxResultVerifier`, a contract deployed on the Ethereum network which can verifiably prove that a transaction was executed on the NEAR network.  You could say this contract is available for use by any other contracts running on the **Ethereum network** as a *Near Transaction Result Oracle*.
+
+Once deployed, `EthEventVerifier` and `NearTxResultVerifier` are available for use by smart contracts operating in their respective environments.
+
+## API
+
+### EthClient
+
+```rust
+pub struct DoubleNodeWithMerkleProof {}
+impl DoubleNodeWithMerkleProof {}
+pub struct HeaderInfo {}
+
+#[near_bindgen]
+pub struct EthBridge {}
+impl Default for EthBridge {}
+
+#[near_bindgen]
+impl EthBridge {
+    pub fn init(validate_ethash: bool, dags_start_epoch: u64, dags_merkle_roots: Vec<H128>) -> Self {}
+    pub fn apply_merkle_proof(&self, index: u64) -> H128 {}
+    pub fn initialized(&self) -> bool {}
+    pub fn last_block_number(&self) -> u64 {}
+    pub fn dag_merkle_root(&self, epoch: u64) -> H128 {}
+    pub fn block_hash(&self, index: u64) -> Option<H256> {}
+    pub fn block_hash_safe(&self, index: u64) -> Option<H256> {}
+    pub fn add_block_header(&mut self, block_header: Vec<u8>, dag_nodes: Vec<DoubleNodeWithMerkleProof>) {}
+}
+```
+
+
+### NearClient
+
+```java
+contract NearBridge is Ownable {
+    mapping(uint256 => bytes32) public blockHashes;
+    function addBlockHeaders(bytes[] memory blockHeaders) public onlyOwner {}
+}
+```
+
+
+### Usage
+
+TODO
+
+
+## Data Structures
+
+### NEAR Blockchain
+
+**Primitives**
+
+*Please refer to [nearcore primitive types](https://github.com/nearprotocol/nearcore/blob/master/core/primitives/src/types.rs) for details*
+
+```
+Gas         u64
+Nonce       u64
+Balance     u128
+AccountId   String
+```
+
+**Crypto**
+
+*Please refer to [nearcore crypto signature](https://github.com/nearprotocol/nearcore/blob/master/core/crypto/src/signature.rs) for details*
+
+```
+struct CryptoHash([u8; 32])                                     Sha256 digest
+struct MerkleHash(CryptoHash) 
+struct EpochId(CryptoHash) 
+
+enum PublicKey
+    ED25519(sodiumoxide::crypto::sign::ed25519::PublicKey)      [u8; 32]
+    SECP256K1(Secp256K1PublicKey)                               [u8; 64]
+
+
+enum Signature
+    ED25519(sodiumoxide::crypto::sign::ed25519::Signature)      [u8; 64]
+    SECP256K1(Secp2561KSignature)                               [u8; 65]
+```
+
+**Block header**
+
+*Please refer to [nearcore block primitives](https://github.com/nearprotocol/nearcore/blob/master/core/primitives/src/block.rs) for details*
+
+```
+struct ValidatorStake
+    account_id              AccountId               Account that stakes money.
+    public_key              PublicKey               Public key of the proposed validator.
+    amount                  Balance                 Stake / weight of the validator.
```

Comment:
Maybe better to keep it Java/C style, like:
```java
struct ValidatorStake {
      AccountId  account_id; // Account that stakes money.
      PublicKey public_key;  // Public key of the proposed validator.
      Balance amount;
}
```

Reply:
✅